### PR TITLE
Adds distributed, semi-durable bindings to lattice. 

### LINKED ISSUE #57: Capability provider bindings must be regenerable
When a capability provider is added to a host and lattice mode is active, then the host will perform a `bindings` inventory probe on the lattice for that provider's ID+binding name pair and re-establish those bindings. This should have the net effect on consumers of appearing as though bindings within a lattice are durable.

Use case / acceptance scenario:
* Start 3 wascc-hosts:
  * Actor A in host A bound to provider C 
  * Actor B in host B bound to provider B
  * Provider B runs in host B
  * Provider C runs in host C
* Invoke actors A and B to verify that baseline functionality works
* Kill host C
* Restart host C without performing any manual configuration
* Invoke actor A, ensuring that once again the baseline functionality works because the binding has been re-established.

This is probably easiest to demonstrate with an HTTP server, where the termination of the provider will shut down the HTTP server and free up the TCP port, and when the provider is brought back up again, the result of probing the lattice will re-establish the binding between the existing actor and provider. We should also be able to observe that no other bindings, actors, or providers are affected by this process.

### LINKED ISSUE #65: Support wasm3 as an alternative underlying engine
This requires the [waPC](https://github.com/wapc/wapc-rust) Rust host SDK to also support the `wasm3` engine as a feature flag.

### LINKED ISSUE #77: Binding via API calls/manifest of an actor to a provider only works in host with actor
This should ultimately work in all of the following scenarios:
* Bind actor to provider via API call in provider's host
* Bind actor to provider via API call in actor's host (**works today**)
* Bind actor to provider via API call in host that holds neither the actor nor the provider

### LINKED ISSUE #83: Rationalize the nature of bindings in the context of the lattice
The original concept of bindings was created for an isolated, single-process system that had no knowledge of a lattice. Right now, the way we're handling bindings isn't quite in accordance with the kind of functionality we want to support in the lattice. There are a number of things that we need to do in order to bring bindings up to par with lattice functionality:

* Rename the host's `bind_actor` method to `set_binding` (this will overwrite any previously existing binding for the triplet of actor-provider-provider_instance_name (e.g. "default")
* Add a `remove_binding` function (there isn't one today, only `remove_actor`)
* **(Implementation Detail)** - All bindings must be _shared knowledge_ among every member of the lattice. This means a couple of things:
    * Each host will _queue subscribe_ to the inventory query to obtain bindings, since that will no longer be a scatter-gather operation. Each host can answer with the complete list of bindings
    * When a binding is created, _from any host_, regardless of where the actor and capability are located, the bus holding the actor in question will subscribe to the private subject for that actor<->capability binding.
    * When a binding is removed, _from any host_, regardless of where the actor and capability are located, the bus holding the actor in question will _unsubscribe_ from the private subject for that actor<->capability binding. 
* When an actor is removed, the bus will unsubscribe from all pertinent actor subjects (including all of its bound providers), but the _binding information will not be removed_
* When a capability provider is removed, all bound actors will unsubscribe from all relevant subjects, but the _binding data will remain_ in the lattice.
* When an actor is added to a host, the bus will be checked for relevant bindings, and the actor will subscribe to all private comms subjects for each discovered binding involving that actor.
* When a capability provider is added to a host, it will receive `OP_BIND_ACTOR` messages for all of its existing bindings, allowing it to provision all appropriate resources and establish communications on the right bus subject. This will close issue #46 

# Acceptance Criteria
To verify that this is working as planned:

* All previous non-lattice tests must continue to work. The host needs to work as it always has when lattice mode isn't enabled.
* With lattice enabled, we should be able to perform the following test:
    * Create 2 hosts with _no_ actors and _no_ capability providers
    * Add an actor to each host
    * Add a provider to each host
    * Call `bind_actor` from either host to establish actor-provider bindings (2 bindings)
    * Remove actor from host 1 (binding should remain in the lattice)
    * Add actor to host 2 (binding should not duplicate, should not be re-initialized by provider, actor should subscribe to right topic, and we can make a call on the provider that will talk to the actor
    * Remove actor 2 from host 2
    * Add actor 2 to host 1
    * Verify actor 2 continues to function normally
    * Remove provider 1 from host 1
    * Add provider 1 to host 2
    * Verify that all of provider 1's bound actors continue to work and that it responds to all of its appropriate stimuli
    * Call `remove_binding` for actor 1-provider 1. Assuming this provider is an HTTP server, after this binding removal, the actor should _still be in the lattice_, but attempting to `curl` the previously bound port should give a `no response` error because the removal of that binding should have shut down the relevant resources.

### PR DESCRIPTION
**Pull request by autodidaddict**
Also adds support for selectable wasm engines

Closes #83, #65 , and #57

### CODE REVIEW

**Comment by autodidaddict** [comment]
Test suites pass locally but aren't working in the GH pipeline. Likely a concurrency issue, I'll try and set the tests to single-threaded.

File: src/bus/lattice.rs [hunk 1]
```diff
@@ -124,13 +135,62 @@ impl DistributedBus {
             backoffcount += 1;
         }
         let _ = self.publish_event(BusEvent::HostStopped(self.host_id.to_string()));
+        std::thread::sleep(Duration::from_millis(300));
```
Comment (brooksmtownsend) [inline src/bus/lattice.rs:138]: Is there a better way for us to know the host has stopped? What if it takes longer than 300 millis for that message to get delivered, etc.

File: src/lib.rs [hunk 2]
```diff
@@ -181,7 +182,8 @@ impl HostBuilder {
     }
 
     /// Sets the lattice namespace for this host. A lattice namespace is a unit of multi-tenant
-    /// isolation on a network
+    /// isolation on a network. To reduce the risk of conflicts or subscription failures, the
+    /// lattice namespace should not include any non-alphanumeric characters.
```
Comment (brooksmtownsend) [inline src/lib.rs:186]: Do we want to enforce this?

File: src/lib.rs [hunk 3]
```diff
@@ -720,7 +774,7 @@ impl Host {
     }
 
     /// Returns the list of actors in the host that contain all of the tags in the
-    /// supplied parameter
+    /// supplied parameter. This function will not make a lattice-wide tag query
```
Comment (brooksmtownsend) [inline src/lib.rs:780]: Theres a similar comment on line 757-758, I don't see it yet but something to think about is if we get a decent amount of functions that differ in their functionality with lattice mode on / off, we may want to find a way to easily distinguish them so there's less ambiguity for the developer of the host

File: src/spawns.rs [hunk 4]
```diff
@@ -149,7 +146,7 @@ pub(crate) fn spawn_actor(
                             }
                             resp_s.send(r).unwrap();
                             continue;
-                        }
+                        }*/
```
Comment (brooksmtownsend) [inline src/spawns.rs]: Can this section be removed or is it going to make a comeback

**Review by brooksmtownsend — COMMENTED** [review]
Just a few comments, https://wascc.dev/tutorials/first-actor/create_host/ (and maybe a few other places in the dev site) will just need a little `s/bind_actor/set_binding/`

File: src/bus/lattice.rs (hunk 1 above)
Comment (autodidaddict) [reply src/bus/lattice.rs:138]: There are some ways to know, but often detecting them requires using the lattice feature when that feature might not be available during the test ;)

File: src/lib.rs (hunk 3 above)
Comment (autodidaddict) [reply src/lib.rs:780]: I ultimately want to get rid of non-lattice mode, and aim for having lattice be the normal behavior, and we can simply choose in or out of process lattice.

File: src/spawns.rs (hunk 4 above)
Comment (autodidaddict) [reply src/spawns.rs]: This won't make a comeback. Issue #79 will ultimately put _something_ back here in its place.

**Comment by autodidaddict** [comment]
> Just a few comments, https://wascc.dev/tutorials/first-actor/create_host/ (and maybe a few other places in the dev site) will just need a little `s/bind_actor/set_binding/`

Absolutely, there's going to be a number of changes that will need to be made to the dev since once 0.13.0 is released. Keep in mind that master != release

File: src/lib.rs (hunk 2 above)
Comment (autodidaddict) [reply src/lib.rs:186]: Yes :)

**Comment by autodidaddict** [comment]
Also closes #77

**Review by brooksmtownsend — APPROVED** [review]

**Pull request merged**